abstio/src/download.rs

```diff
@@ -33,13 +38,31 @@ pub async fn download_bytes<I: AsRef<str>>(url: I) -> Result<Vec<u8>> {
     Ok(bytes)
 }
 
-/// Downloads a file, printing progress to STDOUT. This must be called with a tokio runtime
-/// somewhere.
+/// Download a file from a URL. This must be called with a tokio runtime somewhere. Progress will
+/// be printed to STDOUT.
 pub async fn download_to_file<I1: AsRef<str>, I2: AsRef<str>>(url: I1, path: I2) -> Result<()> {
-    let bytes = download_bytes(url).await?;
+    let (mut tx, rx) = futures_channel::mpsc::channel(1000);
+    print_download_progress(rx);
+    let bytes = download_bytes(url, &mut tx).await?;
     let path = path.as_ref();
     std::fs::create_dir_all(std::path::Path::new(path).parent().unwrap())?;
     let mut file = std::fs::File::create(path)?;
     file.write_all(&bytes)?;
     Ok(())
 }
+
+/// Print download progress to STDOUT. Pass this the receiver, then call download_to_file or
+/// download_bytes with the sender.
+pub fn print_download_progress(mut progress: mpsc::Receiver<String>) {
+    tokio::task::spawn_blocking(move || loop {
+        match progress.try_next() {
+            Ok(Some(msg)) => {
+                abstutil::clear_current_line();
+                print!("{}", msg);
+                std::io::stdout().flush().unwrap();
+            }
+            Ok(None) => break,
+            Err(_) => {}
```

Comment:
To clarify - if the network transfer fails terminally, my assumption is that we'd end up here and loop infinitely.

Reply:
I guess this isn't the network error - but the Err from reading the progress channel, so I'm not actually sure what the error conditions are. 🤔 

Should we at least log or something so that if we do loop it's a little bit more explicable?

Reply:
According to the API, this is a totally normal case. Not sure why they chose `Err` to represent no messages yet. I added a comment